src/xwayland/xwm/surface.rs

```diff
@@ -0,0 +1,878 @@
+        let self_alive = self.state.lock().unwrap().alive;
+        let other_alive = other.state.lock().unwrap().alive;
+        self.window == other.window && self_alive && other_alive
+    }
+}
+
+/// Errors that can happen for operations on an [`X11Surface`]
+#[derive(Debug, thiserror::Error)]
+pub enum X11SurfaceError {
+    /// Error on the underlying X11 Connection
+    #[error(transparent)]
+    Connection(#[from] ConnectionError),
+    /// Operation was unsupported for an override_redirect window
+    #[error("Operation was unsupported for an override_redirect window")]
+    UnsupportedForOverrideRedirect,
+}
+
+/// Window types of [`X11Surface`]s
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[allow(missing_docs)]
+pub enum WmWindowType {
+    DropdownMenu,
+    Dialog,
+    Menu,
+    Notification,
+    Normal,
+    PopupMenu,
+    Splash,
+    Toolbar,
+    Tooltip,
+    Utility,
+}
+
+impl X11Surface {
+    /// Create a new [`X11Surface`] usually handled by an [`X11WM`]
+    ///
+    /// ## Arguments
+    ///
+    /// - `window` X11 window id
+    /// - `override_redirect` set if the X11 window has the override redirect flag set
+    /// - `conn` Weak reference on the X11 connection
+    /// - `atoms` Atoms struct as defined by the [xwm module](super).
+    /// - `geometry` Initial geometry of the window
+    /// - `log` slog logger to be used by this window
+    pub fn new(
+        window: u32,
+        override_redirect: bool,
+        conn: Weak<RustConnection>,
+        atoms: super::Atoms,
+        geometry: Rectangle<i32, Logical>,
+        log: impl Into<Option<::slog::Logger>>,
+    ) -> X11Surface {
+        X11Surface {
+            window,
+            override_redirect,
+            conn,
+            atoms,
+            state: Arc::new(Mutex::new(SharedSurfaceState {
+                alive: true,
+                wl_surface: None,
+                mapped_onto: None,
+                geometry,
+                title: String::from(""),
+                class: String::from(""),
+                instance: String::from(""),
+                protocols: Vec::new(),
+                hints: None,
+                normal_hints: None,
+                transient_for: None,
+                net_state: Vec::new(),
+            })),
+            user_data: Arc::new(UserDataMap::new()),
+            log: crate::slog_or_fallback(log).new(slog::o!("X11 Window" => window)),
+        }
+    }
+
+    /// X11 protocol id of the underlying window
+    pub fn window_id(&self) -> u32 {
+        self.window
+    }
+
+    /// Set the X11 windows as mapped/unmapped affecting its visibility.
+    ///
+    /// It is an error to call this function on override redirect windows
+    pub fn set_mapped(&self, mapped: bool) -> Result<(), X11SurfaceError> {
+        if self.override_redirect {
+            if mapped {
+                return Ok(());
+            } else {
+                return Err(X11SurfaceError::UnsupportedForOverrideRedirect);
```

Comment:
I thought override-redirect windows map themselves and I just get a `map_notify`, so downstream doesn't need to map it? (and shouldn't explicitly?)

Reply:
From the point of view of the WM: Sure.

I am just suggesting one of two changes:

- Change the doc to only say "it is an error to call this on override redirect windows *with argument false*" (making it clear that the function can be called on these function
- Change the implementation to `if self.override_redirect { return Err(X11SurfaceError::UnsupportedForOverrideRedirect); }` so that it does what its doc comment says.

I guess the best option would be to extend the doc comment a bit? Something like "Override redirect windows cannot be unmapped through this function" might be what is intended? I'm not sure.

Reply:
> Change the implementation to if self.override_redirect { return Err(X11SurfaceError::UnsupportedForOverrideRedirect); } so that it does what its doc comment says.

Done. I think this was from a previous iteration, that merged `MappedNotify` and `MapRequest` events until I decided to make it more explicit.
Didn't notice this implementation detail at first. Good call as well.